### PR TITLE
feat: Input component

### LINKED ISSUE #5360: suite-native: Search Input Component
**Design:** [Figma](https://www.figma.com/file/Z6AGVUmKQzLNtDozFamW7f/s2-Mobile?node-id=40%3A1185)
Screenshot 2022-05-09 at 22 23 24

**Spec:** 
1. Recommended props:
```tsx
type InputProps = {
	value: string,
	onChange: (value: string) => void,
	placeholder: React.ReactNode,
	isClearable?: boolean,
}
```
2. Don't forget tests

### PR DESCRIPTION
**Pull request by juriczech**
Screenshots of feature:
- https://tppr.me/f56d7

https://user-images.githubusercontent.com/36101761/169248891-767e67f0-656d-4713-a3fd-829edc60d9c6.mov

Closes [#5360](https://github.com/trezor/trezor-suite/issues/5360)

### CODE REVIEW

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [inline packages/atoms/src/SearchInput.tsx]: use width from `utils.borders`

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [inline packages/atoms/src/SearchInput.tsx]: use `useCallback`

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [inline packages/atoms/src/SearchInput.tsx]: use `useCallback`

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [inline packages/atoms/src/SearchInput.tsx]: avoid unnecessary guards etc. it should be added only when needed

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [inline packages/atoms/src/SearchInput.tsx:60]: use `useCallback`

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [inline packages/atoms/src/SearchInput.tsx]: I think TouchableOpacity will be better choice here

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [inline packages/atoms/src/SearchInput.tsx]: don't use font sizes directly, there should be `...utils.typopography.body`

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [inline packages/atoms/src/SearchInput.tsx]: I not sure but I think `flex: 1` is enough, no need to add height

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [inline packages/atoms/src/SearchInput.tsx:33]: I would prefer to have InputWrapper as standalone component, because we will need exact same wrapper in Input a probably Select too

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [inline packages/atoms/src/SearchInput.tsx]: please keep StyleProps right above style where it is used

File: packages/theme/src/colors.ts
Comment (Nodonisko) [inline packages/theme/src/colors.ts]: this color is not used anywhere? 

also always consult adding theme color with Ondra graphics guy, because we need to keep our palette 1:1 with Figma and he will have to add color, and figure colors for othere theme variants etc. 

one more point, I think we have transparentize function in utils `utils.transparentize(0.05)(utils.colors.red)` should do same job

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: packages/atoms/src/SearchInput.tsx
Comment (juriczech) [reply packages/atoms/src/SearchInput.tsx]: @Nodonisko  What's the purpose of using `useCallback` for this one? Since there is no value that should be memoized, I think it's better to have inline function / normal callback function for this one. Since it doesn't add any value with memoisation, it actually comes with performance cost to use the `useCallback` hook. 

If the argument is to always use the hook so that it's harder to not use it when it should be used (if that makes sense 😃), then I'll switch it. Just wanted to point out that it might not be what we want here.

https://kentcdodds.com/blog/usememo-and-usecallback

File: packages/atoms/src/SearchInput.tsx
Comment (juriczech) [reply packages/atoms/src/SearchInput.tsx]: [8dde2c5](https://github.com/trezor/trezor-suite/pull/5433/commits/8dde2c5f1c1693cc984724aa9c04277bf7047375)

File: packages/atoms/src/SearchInput.tsx
Comment (juriczech) [reply packages/atoms/src/SearchInput.tsx]: [8dde2c5](https://github.com/trezor/trezor-suite/pull/5433/commits/8dde2c5f1c1693cc984724aa9c04277bf7047375)

File: packages/atoms/src/SearchInput.tsx
Comment (juriczech) [reply packages/atoms/src/SearchInput.tsx]: [8dde2c5](https://github.com/trezor/trezor-suite/pull/5433/commits/8dde2c5f1c1693cc984724aa9c04277bf7047375)

File: packages/atoms/src/SearchInput.tsx
Comment (juriczech) [reply packages/atoms/src/SearchInput.tsx]: [8dde2c5](https://github.com/trezor/trezor-suite/pull/5433/commits/8dde2c5f1c1693cc984724aa9c04277bf7047375)

File: packages/atoms/src/SearchInput.tsx
Comment (juriczech) [reply packages/atoms/src/SearchInput.tsx]: [8dde2c5](https://github.com/trezor/trezor-suite/pull/5433/commits/8dde2c5f1c1693cc984724aa9c04277bf7047375)

File: packages/atoms/src/SearchInput.tsx
Comment (juriczech) [reply packages/atoms/src/SearchInput.tsx]: [8dde2c5](https://github.com/trezor/trezor-suite/pull/5433/commits/8dde2c5f1c1693cc984724aa9c04277bf7047375)

File: packages/atoms/src/SearchInput.tsx
Comment (juriczech) [reply packages/atoms/src/SearchInput.tsx:33]: [fd66c65](https://github.com/trezor/trezor-suite/pull/5433/commits/fd66c6563cb2faf0fc3e106e2fd2f34166bbec9b)

File: packages/theme/src/colors.ts
Comment (juriczech) [reply packages/theme/src/colors.ts]: [0585e93](https://github.com/trezor/trezor-suite/pull/5433/commits/0585e93bf57a1592805b2c03261c0fd25665537c)

I used it when I thought I'm doing the form input issues together with this one. I removed it for now and will solve it better in next PR.

File: packages/atoms/src/SearchInput.tsx
Comment (dahaca) [reply packages/atoms/src/SearchInput.tsx]: Maybe the argument here is that `Input` is re-rendering on every value change and that callback is recreated too often? Still, I am not sure that for such simple functions there will be a difference in performance. I usually use `useCallback` for referential equality and complex functions in complex components.

File: packages/atoms/src/InputWrapper.tsx
Comment (dahaca) [inline packages/atoms/src/InputWrapper.tsx]: honestly, I am still getting confused by props being at the top :C And the fact that it's different on desktop and mobile is not very cool imo

File: packages/atoms/src/InputWrapper.tsx
Comment (dahaca) [inline packages/atoms/src/InputWrapper.tsx]: Are those decimal paddings the norm in Native? And why is it even different? 🤔

File: packages/atoms/src/InputWrapper.tsx
Comment (dahaca) [inline packages/atoms/src/InputWrapper.tsx]: While it's true that the search input and others share a lot of similarities, a "wrapper" is not very flexible. I have implemented `InputStyles` for the web and that approach is more suitable for sharing between `Select` and `Input` for example, since on the inside the structure is different but the styles are the same. Search input has a different height, always has the magnifying glass, has no label, can be cleared (which makes little sense in the wrapper) and the component should reflect those differences, not a wrapper. So I would not have a separate wrapper component, since this exact one is unlikely to fit any other inputs from figma and trying to do so is too extra.

**Review by dahaca — CHANGES_REQUESTED** [review]
I would like to add that the clear button should be visible only when an input is not empty. :)

File: packages/atoms/src/InputWrapper.tsx
Comment (Nodonisko) [reply packages/atoms/src/InputWrapper.tsx]: Yea, this wrapper is not what I wanted, please remove both of icons and leave only `children` here, that will make component much more flexible.

File: packages/atoms/src/InputWrapper.tsx
Comment (Nodonisko) [reply packages/atoms/src/InputWrapper.tsx]: It can be used and this is exactly what is in Figma, so I think it's not a issue.

File: packages/atoms/src/InputWrapper.tsx
Comment (Nodonisko) [inline packages/atoms/src/InputWrapper.tsx]: Please never extend ComponentProps by StyleProps. Component props type is something like docs for component so we should keep together as much as possible and not shattered across whole component. 

I am quite of fan of defining it twice in each type because you can think about StyleProps as something that's not related and they are just params of standalone function. In case you don't want to define it twice better do something like this `type InputStyleProps = Pick<InputWrapperProps, 'isFocused'>`

File: packages/atoms/src/InputWrapper.tsx
Comment (Nodonisko) [reply packages/atoms/src/InputWrapper.tsx]: I am really not fan of placing it right before component. In my view PropTypes are kind of public API definition for component and that is exactly what you want to see first when you open component file. If you place it right before component it always in different place somewhere in middle depends of how many styles you component have and how long is your component. Second best place is the end of file, because it's at least always on the of file.

Would be really nice to have this unified but I not willing to do that sacrifice until I will be convinced that there are benefits of having it before component :D

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [reply packages/atoms/src/SearchInput.tsx]: I kind of fan where you use it everywhere, because I never seen any performance degradation because of overusing it, but I have seen lot of performance problems because it was not used.

For now let's keep it simple here and don't do that for these simple components and we will see if it will affect performance.

File: packages/atoms/src/InputWrapper.tsx
Comment (Nodonisko) [reply packages/atoms/src/InputWrapper.tsx]: Actually I spend little bit more time testing it and we will have try different approach. As you can see real TextInput is very small part of this SearchInput component:
Screenshot 2022-05-19 at 21 07 09

That is making UX significantly worse because you have very limited space where you can click and it will focus input. I think correct behaviour is when you click anywhere into input including search icon, it will focus input. To achieve that you will probably need to kill this component and move most of it's styled directly to TextInput component. Also you will probably need to add Pressable to search icon, and focus TextInput when icon is pressed using TextInput ref.

File: packages/atoms/src/InputWrapper.tsx
Comment (Nodonisko) [reply packages/atoms/src/InputWrapper.tsx]: Also pixels in RN are not real physical pixels but something like "relative" pixel, because of scaling.

Comment (juriczech) [reply packages/atoms/src/InputWrapper.tsx]: @Nodonisko @dahaca I have to say I wasn't too sure about this myself. I was considering using the value 16 which is in theming, but then I don't know whether it's better to do it pixel perfect or slight changes depending on theming are to be expected. 

Shall we leave it as it is, or would you prefer to use rounded values always?

File: packages/atoms/src/InputWrapper.tsx
Comment (juriczech) [reply packages/atoms/src/InputWrapper.tsx]: I have to say, that for me it's also better to place prop types to the top of component because sometimes I just want to check file to see what props I need to pass to component and I don't want to look for them under the styles, which shouldn't bother me at that moment. But I see that on suite desktop it is being placed right above the component. 

Maybe I'll put it right above the component and then we can decide on code-improvements whether we should change it or not? It will be better to have it consistent throughout the whole project for people being able to move between packages without having to change to other code styles / component structures.

File: packages/atoms/src/SearchInput.tsx
Comment (juriczech) [reply packages/atoms/src/SearchInput.tsx]: Yeah I think the performance optimization is just way too small to even be noticed. I just added the article to explain my thought process behind choosing not to use the `useCallback` hook. Should we just use it for everything then? Because the performance hit will come in much worse in cases where it's actually helpful and is not used at all. @Nodonisko @dahaca

File: packages/atoms/src/InputWrapper.tsx
Comment (juriczech) [reply packages/atoms/src/InputWrapper.tsx]: [74ed1c1](https://github.com/trezor/trezor-suite/pull/5433/commits/74ed1c1dfb41f1b97d929e2fbfc12f74b6f8c8f6)

File: packages/atoms/src/InputWrapper.tsx
Comment (juriczech) [reply packages/atoms/src/InputWrapper.tsx]: Yeah, good catch! It's fixed now. I moved the wrapper back to the input component and when I do the other input / dropdown component, I will have a look if there are similarities that could be extracted into wrapper that will be more flexible. 

Added TouchableWithoutFeedback to wrap the whole input component which should take care of focusing the input.

File: packages/atoms/src/InputWrapper.tsx
Comment (juriczech) [reply packages/atoms/src/InputWrapper.tsx]: Fixed

File: packages/atoms/src/InputWrapper.tsx
Comment (juriczech) [reply packages/atoms/src/InputWrapper.tsx]: Fixed

**Comment by juriczech** [comment]
> I would like to add that the clear button should be visible only when an input is not empty. :)

Good catch. Fixed now.

File: packages/atoms/src/InputWrapper.tsx
Comment (Nodonisko) [reply packages/atoms/src/InputWrapper.tsx]: Now we have one file that is different from all others in package. Put it back to the top pls, at least keep consistency per package which is fine.

File: packages/atoms/src/InputWrapper.tsx
Comment (Nodonisko) [reply packages/atoms/src/InputWrapper.tsx]: It's fine, we can change it if it will become a problem

File: packages/atoms/src/InputWrapper.tsx
Comment (juriczech) [reply packages/atoms/src/InputWrapper.tsx]: [5895f89](https://github.com/trezor/trezor-suite/pull/5433/commits/5895f89b726e8cc07e9f5fd0be96d7fc97bb8c71)

**Review by Nodonisko — APPROVED** [review]

File: packages/atoms/src/SearchInput.tsx
Comment (dahaca) [reply packages/atoms/src/SearchInput.tsx]: I heard other people using it everywhere too. Honestly, I can't say what is better: one is easier but bight cause problems the other one is more perfectionist and performant but requires brain work :D Might be a topic for code improvements + let's scour some open source React repos and see what they are up to :] @Nodonisko

File: packages/atoms/src/InputWrapper.tsx
Comment (dahaca) [reply packages/atoms/src/InputWrapper.tsx]: Can you just add paddings to the input? So that it's all touchable (except for the close button) and the icon is below? Idk if that absolute positioning is possible in RN. But `TouchableWithoutFeedback` makes sense to me :)

Comment (dahaca) [reply packages/atoms/src/InputWrapper.tsx]: And also what about having input styles separately like on desktop?

File: packages/atoms/src/SearchInput.tsx
Comment (juriczech) [reply packages/atoms/src/SearchInput.tsx]: @dahaca Do you agree I just leave it like this so it doesn't block this PR and then if we decide to have it consistently throughout the whole repo, I will take care of refactoring it? So far it's less then 10 files and it's just moving the position of types.

Comment (dahaca) [reply packages/atoms/src/SearchInput.tsx]: Yes! I unresolved the thread just for visibility :)

File: packages/atoms/src/InputWrapper.tsx
Comment (Nodonisko) [reply packages/atoms/src/InputWrapper.tsx]: Separating some core styles or maybe creating some BaseInput component makes sense, but I think it will be better to do that when we will going to actually create Input component, let's keep this PR simple.

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [inline packages/atoms/src/SearchInput.tsx]: we don't need memo for such a simple operation

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [reply packages/atoms/src/SearchInput.tsx]: nit: I think maybe name `isClearButtonVisible` could be slightly better name, but I will leave it up to you

Comment (juriczech) [reply packages/atoms/src/SearchInput.tsx]: fixed

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [inline packages/atoms/src/SearchInput.tsx]: please remove this comment, it's saying thing that is obvious

File: packages/atoms/src/SearchInput.tsx
Comment (Nodonisko) [inline packages/atoms/src/SearchInput.tsx]: nit: var declartions should be placed before handlers

**Review by Nodonisko — APPROVED** [review]

File: packages/atoms/src/InputWrapper.tsx
Comment (Nodonisko) [reply packages/atoms/src/InputWrapper.tsx]: I created point for code improvements, we can discuss it there

**Review by bouchja1 — APPROVED** [review]

File: packages/atoms/src/SearchInput.tsx
Comment (juriczech) [reply packages/atoms/src/SearchInput.tsx]: Fixed

File: packages/atoms/src/SearchInput.tsx
Comment (juriczech) [reply packages/atoms/src/SearchInput.tsx]: Fixed

**Review by Nodonisko — APPROVED** [review]

**Review by dahaca — APPROVED** [review]

**Pull request merged**
